[docs] Flux group offload (#10847)
huggingface/diffusers: docs/source/en/api/pipelines/flux.md

    <figcaption class="mt-2 text-sm text-center text-gray-500">IP-Adapter examples with prompt "wearing sunglasses"</figcaption>
</div>

## Optimize

> Flux is a very large model and requires ~50GB of RAM. Enable some of the optimizations below to lower the memory requirements.

Reviewer: nice!, but the 50 GB of RAM are used when using group offloading not before, also @a-r-r-o-w was going to check if this is the real number or not, I get this but maybe there's something in my env that makes it go that high. It should use around 20GB for the transformer model in theory.

Author: Ah ok, I'll update this number once we get a clearer value from @a-r-r-o-w!

> Flux is a very large model and requires ~50GB of RAM. Enable some of the optimizations below to lower the memory requirements.

Reviewer 2: ```suggestion
Flux is a very large model and requires ~50GB of RAM/VRAM to load all modeling components. Enable some of the optimizations below to lower the memory requirements.
```

### Group offloading

> [Group offloading](../../optimization/memory#group-offloading) saves memory by offloading groups of internal layers rather than the whole model or weights. Use [`~hooks.apply_group_offloading`] on a model and you can optionally specify the `offload_type`. Setting it to `leaf_level` offloads the lowest leaf-level parameters to the CPU instead of offloading at the module-level.

Reviewer 2: ```suggestion
[Group offloading](../../optimization/memory#group-offloading) lowers VRAM usage by offloading groups of internal layers rather than the whole model or weights. Use [`~hooks.apply_group_offloading`] on a model and you can optionally specify the `offload_type`. Setting it to `leaf_level` offloads the lowest leaf-level parameters to the CPU instead of offloading at the module-level.
```

```py
import torch
from diffusers import FluxPipeline
from diffusers.hooks import apply_group_offloading

model_id = "black-forest-labs/FLUX.1-dev"
dtype = torch.bfloat16
pipe = FluxPipeline.from_pretrained(
	model_id,
	torch_dtype=dtype,
)

> apply_group_offloading(

Reviewer 2: Group offloading in itself is not better than sequential cpu offloading from accelerate. Both are slow but I'm not sure why I got sequential offload time numbers to be much higher than group leaf-level. They are the same in theory so I will cc @SunMarc to help take a look as well

Reviewer 2: If we're showcasing group offloading here, I think there's a strong case to show the usage of `use_stream=True`, or mention it somewhere. It speeds up inference quite a lot!

    pipe.transformer,
    offload_type="leaf_level",
    offload_device=torch.device("cpu"),
    onload_device=torch.device("cuda"),
)
apply_group_offloading(
    pipe.text_encoder, 
    offload_device=torch.device("cpu"),
    onload_device=torch.device("cuda"),
    offload_type="leaf_level"
)
apply_group_offloading(
    pipe.text_encoder_2, 
    offload_device=torch.device("cpu"),
    onload_device=torch.device("cuda"),
    offload_type="leaf_level"
)
apply_group_offloading(
    pipe.vae, 
    offload_device=torch.device("cpu"),
    onload_device=torch.device("cuda"),
    offload_type="leaf_level"
)

prompt="A cat wearing sunglasses and working as a lifeguard at pool."

generator = torch.Generator().manual_seed(181201)

> image = pipe(
>     prompt,
> 	width=576,
> 	height=1024,
> 	num_inference_steps=30,
>     generator=generator

Reviewer 2: ```suggestion
image = pipe(
    prompt,
	  width=576,
	  height=1024,
	  num_inference_steps=30,
    generator=generator
```

    <figcaption class="mt-2 text-sm text-center text-gray-500">IP-Adapter examples with prompt "wearing sunglasses"</figcaption>
</div>

## Optimize

Flux is a very large model and requires ~50GB of RAM/VRAM to load all the modeling components. Enable some of the optimizations below to lower the memory requirements.

> ### Group offloading

Reviewer 3: @stevhliu it might be a good idea to consider making it more generally available to all major pipelines with high usage.
